### PR TITLE
Add `CommandExt` for working with `std::process::Command` output streams

### PR DESCRIPTION
**Pull request by Malax**
While working on the Gradle buildpack, I had the need to prefix `gradlew` output so the user can clearly distinguish between `gradlew` and buildpack output. While implementing a solution, @dzuelke mentioned that he has the same need and also some additional ones. So I went the extra mile, made my solution generic and added docs so that it is easy to use. Implementing this is less trivial than it might sound so I thought it's a great fit for `libherokubuildpack` (or whatever it will be called in the future)

It all revolves around a new trait: `CommandExt` which extends `std::process:Command` with (currently) two helpers that allow writing the unbuffered output of `stdout` and `stderr` to a `std::io::Writer`.

This is the main hook to add more complex logic like `tee`ing output to two different locations, prefixing lines and live-modifying streams using standard `String` functions. These helpers are contained in the new `write` module which has the aforementioned helpers. These can be used without `CommentExt` and are standalone. Their only interface is `std::io::Write`.

Here is an example how this can be used:

```rust
use libherokubuildpack::command::CommandExt;
use libherokubuildpack::write::tee;
use std::fs;
use std::process::Command;

let logfile = fs::File::open("log.txt").unwrap();

let exit_status = Command::new("date")
    .spawn_and_write_streams(tee(std::io::stdout(), &logfile), std::io::stderr())
    .and_then(|mut child| child.wait())
    .unwrap();
```

The exposed API has proper rust docstrings. For detailed info and more usage examples refer to those.

Closes GUS-W-12188139

### CODE REVIEW

File: libherokubuildpack/src/command.rs
Comment (schneems) [inline libherokubuildpack/src/command.rs:72]: How does this not violate orphan rules? I thought we couldn't implement a trait for a struct we don't own (broad strokes).

File: libherokubuildpack/src/command.rs
Comment (schneems) [inline libherokubuildpack/src/command.rs:132]: I basically have no clue what's happening here. I'm not familiar with crossbeam. I'm aware it's a parallelism-ish library. 

To write this out it looks like we're using crossbeam scope which allows us to spawn scoped threads  https://docs.rs/crossbeam/0.7.3/crossbeam/fn.scope.html. We're moving the memory of stdou and stderr into those threads and essentially just copying whatever goes into stdout into whatever is passed in to the stdout_writer. I'm assuming this is blocking (inside of the thread).

Then we do some stuff which essentially boils down to joining the threads.

I'm not entirely following how this last line works, but it compiles so i'm sure it does:

```rust
        // Return the first error from either Result or the child process value
        stdout_copy_result.and(stderr_copy_result).map(|_| child)
```

Comment (schneems) [reply libherokubuildpack/src/command.rs:132]: Notes for self: My naieve implementation is here https://github.com/heroku/buildpacks-ruby/blob/f33fd189414507645d563ceea675157c78847fc5/commons/src/env_command.rs#L330-L376. I used channels to get data out of the threads without lifetime issues.

**Review by schneems — COMMENTED** [review]
Still working on the review. Wanted to leave these questions/notes for you. I'll pick up the write.rs and finish up later.

Overall looks good. I like having some of this general purpose goodness in shared libraries.

File: libherokubuildpack/src/command.rs
Comment (Malax) [reply libherokubuildpack/src/command.rs:72]: Since this crate defines the `CommandExt` trait it's fine. The crate implementing a trait must either define the trait _or_ the type it's implemented for.

File: libherokubuildpack/src/command.rs
Comment (Malax) [reply libherokubuildpack/src/command.rs:132]:
```
// Return the first error from either Result or the child process value
stdout_copy_result.and(stderr_copy_result).map(|_| child)
```

The [`and`](https://doc.rust-lang.org/std/result/enum.Result.html#method.and) does the following:
- If `stdout_copy_result` is an `Err`, return it and do nothing with `stderr_copy_result`.
- If `stdout_copy_result` is an `Ok`, return `stderr_copy_result` as is (so either an `Err` or an `Ok`).
- This means, after the `and`, we will have either the `Err` of `stdout_copy_result` or the `Ok` or `Err` of `stderr_copy_result`.

The `map` is a regular map, but in this case we actually don't mind what's in the `Ok`. If the `Result` after the `and` is `Ok` (meaning that both succeeded), we put `child` into the `Ok`, dropping whatever was in the `Ok` previously.

Comment (schneems) [reply libherokubuildpack/src/command.rs:132]: > With crossbeam, we can convince the compiler to believe us that OW and EW are only referenced by the threads until the crossbeam scope ends.

That was my understanding as well. I think it's commented well enough. I'm just not familiar with the API so it was more of a talk-aloud. If my comments are generally correct there's nothing to update. Thanks for taking the time to add those annotations

File: libherokubuildpack/src/command.rs
Comment (schneems) [inline libherokubuildpack/src/command.rs:65]: I don't love these method names. I know they match Command's but I find them confusing. 

Looking at the code both `output` and  `spawn` seem like generic actions rather than behaviors.

Would you be open to `capture_output_and_write_streams` or `capture_and_write_streams`. Maybe just `capture_and_stream`.

**Review by schneems — APPROVED** [review]
Added some additional comments. 

- One question about whether we need Clone or why we might not want it. 
- Another comment related to naming. I would like to add `capture` to `output_and_write_streams` or otherwise indicate what it's doing as I feel the existing naming of `spawn` or `output` isn't very descriptive. I would prefer it but don't want to block on it if you feel strongly otherwise.

File: libherokubuildpack/src/write.rs
Comment (Malax) [reply libherokubuildpack/src/write.rs:33]: Not sure what you mean by this comment. Yeah, it does with with both `\n` and `\r\n` so we're good when we think about Windows and *NIX. The Rustdoc goes a bit into detail, explaining what it does under the hood. If there is some other ecosystem where lines are separated by some other delimiter this won't work, correct. 

Do you want me to change the name of this function to smth. like `newline_delimited_map`?

File: libherokubuildpack/src/write.rs
Comment (Malax) [reply libherokubuildpack/src/write.rs:49]: I guess YAGNI is the way to go here - if there are no obvious benefits of implementing a specific trait (either manually or via `derive`) I would wait until we need it so we a) don't add things we don't need and b) know _how_ we should implement them. When thinking about `Debug` for example, what do we want, how should we represent the function? Do we need to?

File: libherokubuildpack/src/write.rs
Comment (Malax) [reply libherokubuildpack/src/write.rs:182]: `output_and_write_streams` would contain the _original_ output. The manipulations via the `Write`s happens separately and _in addition_. This ensures `output_and_write_streams` is a drop-in replacement for `output`. `CommandExt` shouldn't really be a concern of this module, it only contains `Write` implementations that can be used wherever. I use `CommandExt` in the Rustdoc as an example because that is what they are designed for, but the implementation itself shouldn't be coupled to it.

If you want to capture the modified stream, a `&mut [u8]` is `Write`. So you can easily "pipe" it into a variable of that type and inspect it after the command finished.

File: libherokubuildpack/src/command.rs
Comment (Malax) [reply libherokubuildpack/src/command.rs:65]: Hm, I'm quite happy with the naming of these - especially compared to earlier iterations I had. 😂 

I think it's beneficial to name them after the `Command` functions they emulate (they are drop-in replacements). I think of these functions and an additional "tap" for the output streams - in the form of the `Write` interface. 

Do you think that `Command::spawn` and `Command::output` should be renamed to `Command::capture` and `Command::capture_output` too? I think at least `capture` is a bit misleading. In any case, I don't see the benefit of inventing another name for the same thing that users then have to lean in addition - no matter how "bad" the original naming is.

File: libherokubuildpack/src/write.rs
Comment (schneems) [reply libherokubuildpack/src/write.rs:33]: I'm pushing up my glasses and squinting while thinking "well actually" in a nasally voice while writing the original comment. I think we shouldn't need to change the name `line_mapped` makes sense. If we wanted to make a change I think specifying "newline" in the rustdoc would be the way to go.

File: libherokubuildpack/src/command.rs
Comment (schneems) [reply libherokubuildpack/src/command.rs:65]: TLDR; I agree but it took awhile to get there

> Do you think that Command::spawn and Command::output should be renamed to Command::capture and Command::capture_output too? 

I care about behaviors `Command::capture_output` would be better than `Command::output` imho. I guess you could advocate for `Command::output` being named because it GIVES you the output. But if that's the case `Command::spawn` makes no sense because it does not GIVE you the `spawn` (what does that even mean?). It gives you a child which I guess is a spawned process https://doc.rust-lang.org/std/process/struct.Child.html. So, meh...maybe that makes sense. But I've been using `Command` for a long time and it took me until now to make that connection. 

I think that it __technically__ makes sense, but only when you squint. I have a comparison to these functions: 

https://github.com/heroku/buildpacks-ruby/blob/schneems/fresh-start-rust/commons/src/in_app_dir_cache.rs

While technically `copy_app_path_to_cache` and `move_app_path_to_cache` are correctly named but the subtle difference is easy to miss, so I renamed one of them `destructive_move_app_path_to_cache` so it is PAINFULLY obvious what you're getting it when reading it.

So i think to sum up my argument: `output` and `spawn` might be technically correct and an okay name, but I don't think they communicate behavior or intent to someone coming from a different language or context. I feel that command is actually too low level of a primitive (hence why I wrote my own wrapper struct that none of you like). 

> I think at least capture is a bit misleading.

Can you say more here?

> I don't see the benefit of inventing another name for the same thing that users then have to learn in addition - no matter how "bad" the original naming is.

I'm thinking about it and using my own arguments...I agree. The difference between capturing and not capturing would be detected at code time as the developer would realize they can't access the stdout/stderr. Versus in the stream/no-stream case the difference to customers is huge. However in both our methods we are streaming and it explicitly says so in the method name `*_and_write_streams`

I'm leaving this long comment, because it's how I derived this conclusion and want to leave it as a note to myself.

File: libherokubuildpack/src/write.rs
Comment (edmorley) [reply libherokubuildpack/src/write.rs:49]: > So in general I think we should at least consider implementing these derives by default if we can: Clone, Default, PartialEq, PartialOrd, Hash, Eq, and Ord. Less because we'll directly need those things on this struct, and more that any struct that contains this one is blocked from deriving these traits if it contains our struct.

Some thoughts:
- In the general case, yes we should at least consider whether some of these are useful when adding new code.
- `and more that any struct that contains this one is blocked from deriving` -> good point, though in this specific case, why would they need to put the mapper in their own struct?
- Adding derives isn't free - it has an impact on compile times (there's been a bunch of performance work, eg [here](https://nnethercote.github.io/2022/07/20/how-to-speed-up-the-rust-compiler-in-july-2022.html), trying to reduce the size of the generated code, since in some cases it's quite verbose and gives the compiler extra work to do). This shouldn't stop us using them if there is a genuine chance someone will need them, but we should probably not default to adding all 7 of those derive attributes to every struct without thinking about it at least briefly case by case.

File: libherokubuildpack/src/write.rs
Comment (edmorley) [inline libherokubuildpack/src/write.rs:6]:
```suggestion
/// Constructs a writer that buffers written data until the given marker byte is encountered and
/// then applies the given mapping function to the data before passing the result to the wrapped
```

File: libherokubuildpack/Cargo.toml
Comment (edmorley) [inline libherokubuildpack/Cargo.toml:39]: The `crossbeam` crate itself is just a re-export of half a dozen other crossbeam crates:
https://github.com/crossbeam-rs/crossbeam#crates

It looks like the `crossbeam` features used in this PR are from the `crossbeam-utils` crate only.

Using that crate standalone seems to be a supported/expected use-case, given the docs:
https://crates.io/crates/crossbeam-utils

Please could we switch to `crossbeam-utils` for now, to reduce the number of dependencies that are downloaded and compiled but never needed? We can always switch back to the `crossbeam` crate if we ever start using several of the `crossbeam-*` crates :-)

Also pedantic nit: The dependencies list was previously alphabetical, now it's not 😆

**Review by edmorley — COMMENTED** [review]
Thank you for adding this!

File: libherokubuildpack/Cargo.toml
Comment (Malax) [reply libherokubuildpack/Cargo.toml]: Changed in [49ac7ea](https://github.com/heroku/libcnb.rs/pull/535/commits/49ac7ea4752b57a81d908cef478f230d109834fe)

File: libherokubuildpack/src/write.rs
Comment (Malax) [reply libherokubuildpack/src/write.rs:33]: Isn't this already in the rustdocs? Or do you mean something else?
> Constructs a writer that buffers written data until an ASCII/UTF-8 newline byte (`0x0A`)

File: libherokubuildpack/src/command.rs
Comment (Malax) [reply libherokubuildpack/src/command.rs:65]: I'm marking this is "resolved", LMK if I misunderstood your comment and feel free to re-open! :)

**Review by edmorley — APPROVED** [review]

File: libherokubuildpack/src/write.rs
Comment (Malax) [reply libherokubuildpack/src/write.rs:51]: Looking at the standard library for similar types that serve as intermediate types (often named after the function that creates them: `FlatMap`, `Child`, etc.), I changed my mind and think `Clone` and `Debug` make sense here. Except for `Debug` in the case of `MappedWrite`, they can be automatically derived and the `Debug` implementation isn't too bad.

I agree with the "lost momentum" argument. :)

Thanks for raising this! Added in 86462bee1558713fc4f757d1ad2240b4a6ce4248.

Comment (schneems) [reply libherokubuildpack/src/write.rs:51]: Thank you for this @Malax ❤️ . 

> Adding derives isn't free

I meant in terms of programmer/maintainer time. But I hadn't really considered binary size or compile time. In my head I'm still basically laughing at how much faster even "bad" rust code is compared to carefully written Ruby code. It would be nice to get a sense of the cost of that tradoff (size and compile time). I'm not sure how we could do it.

Maybe we could remove deriving Debug from all structs in libcnb, clear the caches and time the compile time and check binary size? For me, i'm in a mode of usually wanting to optimize programmer productivity and experience first and then really only caring about perf and size second. I care a lot about performance in Ruby land mostly as a factor of it ultimately affecting productivity and experience. I'm curious to brainstorm ways to better tune intuition on these tradeoffs.

I think it's a good point, I want to know more. I always want to understand the impact of my (coding) actions.

> why would they need to put the mapper in their own struct?

More coming from a place of assuming people will build structs (and enums) from pretty much any other struct that exists. In this case, I've stored a bunch of stuff for command execution in a struct in my buildpack.

If there's a good reason to NOT allow someone to use or store a struct in their own code, we should probably annotate/document that case (or make the structs private all together). Otherwise as library implementers we need to write for what people might likely do, even if we've not got concrete ideas for why. In this case I don't think leaving out Clone or Debug would make someone second guess their code "hmm, maybe I shouldn't store this in a struct/enum" and instead they would likely jump straight to opening an issue or a PR. 

In general it's good to be opinionated and also have a good sense for what we want to encourage people to do. I think we also need to balance that with some general guidelines about reasonable expectations of what our users will and won't reasonably do.

Comment (edmorley) [reply libherokubuildpack/src/write.rs:51]: I'm not talking about not adding `Debug` (of all the traits, that seems like a no-brainer to add; my post was about the "7 traits" aspect, per: "we should probably not default to adding all 7 of those derive attributes to every struct without thinking about it at least briefly case by case"), nor am I talking about binary size or runtime performance (I only mention compile times). 

I know it's not intentional, but it feels like the response above is nearing straw man territory? :-)

**Pull request merged**
